contracts/Controller.sol

```diff
@@ -122,57 +137,157 @@ contract Controller is Ownable {
+                    otoken.strikeAsset(),
+                    otoken.collateralAsset(),
+                    otoken.expiryTimestamp()
+                )
+            );
+    }
 
     /**
-     * @notice Checks if the sender is the operator of the owner’s account
-     * @param _owner The owner of the account
-     * @param _operator account operator
-     * @return true if it is an account operator, otherwise false
+     * @notice get number of vaults in a specific account
+     * @param _accountOwner account owner address
+     * @return number of vaults
      */
-    // function isOperator(address _owner, address _operator) public view returns (bool) {
-    // }
+    function getAccountVaultCounter(address _accountOwner) external view returns (uint256) {
+        return accountVaultCounter[_accountOwner];
+    }
+
+    /**
+     * @notice function to check if otoken is expired
+     * @param _otoken otoken address
+     * @return true if otoken is expired, else return false
+     */
+    function isExpired(address _otoken) public view returns (bool) {
+        uint256 otokenExpiryTimestamp = OtokenInterface(_otoken).expiryTimestamp();
+
+        return now > otokenExpiryTimestamp;
+    }
+
+    /**
+     * @notice Return a specific vault.
+     * @param _owner account owner.
+     * @param _vaultId vault.
+     * @return Vault struct
+     */
+    function getVault(address _owner, uint256 _vaultId) public view returns (MarginAccount.Vault memory) {
+        return vaults[_owner][_vaultId];
+    }
 
     /**
      * @notice Execute actions on a certain vault
      * @dev For each action in the action Array, run the corresponding action
      * @param _actions An array of type Actions.ActionArgs[] which expresses which actions the user want to execute.
      * @return Vault strcut. The new vault data that has been modified (or null vault if no action affected any vault)
      */
-    // function _runActions(Actions.ActionArgs[] memory _actions) internal returns (MarginAccount.Vault memory) {
-    // }
+    function _runActions(Actions.ActionArgs[] memory _actions) internal returns (MarginAccount.Vault memory) {
+        MarginAccount.Vault memory vault;
+
+        for (uint256 i = 0; i < _actions.length; i++) {
+            Actions.ActionArgs memory action = _actions[i];
+            Actions.ActionType actionType = action.actionType;
+
+            uint256 actionVaultId;
+            bool isActionVaultStored;
+
+            if (actionType == Actions.ActionType.OpenVault) {
+                // check if this action is manipulating the same vault as all other actions, other than SettleVault
+                (actionVaultId, isActionVaultStored) = _checkActionVault(
+                    actionVaultId,
+                    action.vaultId,
+                    isActionVaultStored
+                );
+                _openVault(Actions._parseOpenVaultArgs(action));
+            }
+        }
+
+        return vault;
+    }
```

Comment:
yeah because in this case it is just the open vault action...
`_runActions` should return the vault to pass it to calculator for `isValideState`, even if there is not modified state, this function should return something

Reply:
why can't the `operate` function get the vault again on its own after you run `_runActions`?

Reply:
what's the benefits of that, if there is any ?

Reply:
I think we should keep it consistent that don't return value for non-view functions.
The only function that calls runActions is `operate`, and it can get this variable on its own, so why do we need the runActions to return it?

Reply:
I don't remember the reason behind the return actually, following spec. But yeah I agree getting that at the end once better than assigning it at the end of every action and returning that.